### PR TITLE
feat: voting_escrow contract

### PR DESCRIPTION
**Pull request by epanchee**

### CODE REVIEW

**Comment by wedancedalot** [comment]
we use underscore naming for contracts inside astroport-core repo, also builder_unlock contract is named using underscore in this repo, let's stick to this type of naming

File: contracts/astro-voting-escrow/src/utils.rs [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+};
+use cw_storage_plus::{Bound, U64Key};
+use std::convert::TryInto;
+
+use crate::state::{Point, BLACKLIST, CONFIG, HISTORY, SLOPE_CHANGES};
+
+/// # Description
+/// Checks the time is within limits
+pub(crate) fn time_limits_check(time: u64) -> Result<(), ContractError> {
+    if !(WEEK..=MAX_LOCK_TIME).contains(&time) {
```
[the remark this answers is not in the document]
Comment (epanchee) [reply contracts/astro-voting-escrow/src/utils.rs]: It was clippy's [advice](https://rust-lang.github.io/rust-clippy/master/#manual_range_contains).

[the remark this answers is not in the document]
Comment (ShadoySV) [reply contracts/astro-voting-escrow/src/utils.rs]: It doesn't generate an array: https://doc.rust-lang.org/src/core/ops/range.rs.html#341-376

File: contracts/astro-voting-escrow/src/contract.rs [hunk 2]
```diff
@@ -0,0 +1,667 @@
...
+
+    // disabling ability to extend lock time by less than a week
+    time_limits_check(time)?;
+
+    if lock.end <= get_period(env.block.time.seconds()) {
+        return Err(ContractError::LockExpired {});
+    };
+
+    // should not exceed MAX_LOCK_TIME
+    time_limits_check(lock.end * WEEK + time - env.block.time.seconds())?;
```
[the remark this answers is not in the document]
Comment (epanchee) [reply contracts/astro-voting-escrow/src/contract.rs]: `get_period(time)` just calculates how many periods 'time' contains. It does not calculate what period will be in 'time' seconds. Setting lock.end to get_period(time) will make lock.end in the past.

File: contracts/astro-voting-escrow/src/contract.rs [hunk 3]
```diff
@@ -0,0 +1,667 @@
...
+/// Receives a message of type [`Cw20ReceiveMsg`] and processes it depending on the received template.
+/// If the template is not found in the received message, then an [`ContractError`] is returned,
+/// otherwise returns the [`Response`] with the specified attributes if the operation was successful
+fn receive_cw20(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    cw20_msg: Cw20ReceiveMsg,
+) -> Result<Response, ContractError> {
+    match from_binary(&cw20_msg.msg)? {
```
[the remark this answers is not in the document]
Comment (epanchee) [reply contracts/astro-voting-escrow/src/contract.rs]: Good advice. Fixed

File: contracts/astro-voting-escrow/src/utils.rs [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    addr: &Addr,
+    period_key: &U64Key,
+) -> StdResult<Option<Pair<Point>>> {
+    HISTORY
+        .prefix(addr.clone())
+        .range(
+            deps.storage,
+            None,
+            Some(Bound::Inclusive(period_key.wrapped.clone())),
+            Order::Ascending,
```
[the remark this answers is not in the document]
Comment (epanchee) [reply contracts/astro-voting-escrow/src/utils.rs]: That's true. Fixed.

File: contracts/astro-voting-escrow/src/contract.rs [hunk 5]
```diff
@@ -0,0 +1,667 @@
...
+    remove_addrs: Option<Vec<String>>,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    // Permission check
+    if info.sender != config.owner {
+        return Err(ContractError::Unauthorized {});
+    }
+    let append_addrs = append_addrs.unwrap_or_default();
+    let remove_addrs = remove_addrs.unwrap_or_default();
+    let append = validate_addresses(deps.as_ref(), &append_addrs)?;
```
[the remark this answers is not in the document]
Comment (epanchee) [reply contracts/astro-voting-escrow/src/contract.rs]: Done

File: contracts/astro-voting-escrow/src/contract.rs [hunk 6]
```diff
@@ -0,0 +1,667 @@
...
+    Ok(Response::default().add_attribute("action", "deposit_for"))
+}
+
+/// ## Description
+/// Withdraws whole amount of locked xASTRO.
+/// If lock doesn't exist or it has not yet expired, then an [`ContractError`] is returned,
+/// otherwise returns the [`Response`] with the specified attributes if the operation was successful
+fn withdraw(deps: DepsMut, env: Env, info: MessageInfo) -> Result<Response, ContractError> {
+    let sender = info.sender;
+    blacklist_check(deps.as_ref(), &sender)?;
```
[the remark this answers is not in the document]
Comment (epanchee) [reply contracts/astro-voting-escrow/src/contract.rs]: The poor guy is now allowed to withdraw his funds. Fixed

File: contracts/astro-voting-escrow/src/contract.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ShadoySV) [reply contracts/astro-voting-escrow/src/contract.rs]: There isn't the voting power recalculation with a new coefficient after extending

[the remark this answers is not in the document]
Comment (epanchee) [reply contracts/astro-voting-escrow/src/contract.rs]: It's already included in slope

[the remark this answers is not in the document]
Comment (epanchee) [reply contracts/astro-voting-escrow/src/contract.rs]: Yes, it's true. Fixed this so boost is applied to current voting power if we extend lock's time.

**Review by ShadoySV — APPROVED** [review]

**Pull request merged**